udf/README.md

```diff
@@ -0,0 +1,236 @@
+3. **Anti-Pattern Detection:** When you call the BigQuery UDF, it sends your SQL query to the Cloud Run service, which analyzes it for anti-patterns. The result is returned in JSON format.
+
+
+![bigquery external antipattern function](udf_diagram.png)
+
+## Costs
+
+This tutorial uses billable components of Google Cloud, including the following:
+
+* [BigQuery](https://cloud.google.com/bigquery/pricing)
+* [Cloud Build](https://cloud.google.com/build/pricing)
+* [Cloud Run](https://cloud.google.com/run/pricing)
+
+Use the [pricing calculator](https://cloud.google.com/products/calculator) to generate a cost estimate based on your
+projected usage.
+
+## Before you begin
+
+For this tutorial, you need a Google Cloud [project](https://cloud.google.com/resource-manager/docs/cloud-platform-resource-hierarchy#projects).
+
+1.  [Create a Google Cloud project](https://console.cloud.google.com/projectselector2/home/dashboard).
+1.  Make sure that [billing is enabled](https://support.google.com/cloud/answer/6293499#enable-billing) for your Google
+    Cloud project.
+1.  [Open Cloud Shell](https://console.cloud.google.com/?cloudshell=true).
+
+    At the bottom of the Cloud Console, a [Cloud Shell](https://cloud.google.com/shell/docs/features) session opens and
+    displays a command-line prompt. Cloud Shell is a shell environment with the Cloud SDK already installed, including
+    the [gcloud](https://cloud.google.com/sdk/gcloud/) command-line tool, and with values already set for your current
+    project. It can take a few seconds for the session to initialize.
+
+1.  In Cloud Shell, clone the source repository:
+
+        git clone https://github.com/GoogleCloudPlatform/bigquery-antipattern-recognition.git
+        cd bigquery-antipattern-recognition
+
+2.  Enable all the required Google Cloud APIs
+
+    ```shell
+    gcloud services enable \
+    artifactregistry.googleapis.com \
+    bigquery.googleapis.com \
+    bigqueryconnection.googleapis.com \
+    cloudbuild.googleapis.com \
+    run.googleapis.com \
+    ```
+
+## Deployment script with Terraform
+
+1.  Authenticate using User [Application Default Credentials ("ADCs")](https://cloud.google.com/sdk/gcloud/reference/auth/application-default) as a primary authentication method.
+    ```shell
+    gcloud auth application-default login
+    ```
+
+2.  Initialize and run the Terraform script to create all resources:
+
+    ```shell
+    cd /udf/terraform && \
+    terraform init && \
+    terraform apply
+    ```
+
+## Deployment with Bash script
+
+1.  Authenticate using User [Application Default Credentials ("ADCs")](https://cloud.google.com/sdk/gcloud/reference/auth/application-default) as a primary authentication method.
+    ```shell
+    gcloud auth application-default login
+    ```
+2.  Navigate to `udf/deploy_udf.sh` and modify your variable names for resources:
```

Comment:
Since this is meant to be executed in the cloud shell, navigating to the file and editing requires the use of a command line text editor (e.g. nano). Since this is not the most user friendly experience, cant we have these be env variables. 

```
export PROJECT_ID="<PROJECT_ID>"
export REGION="<REGION_ID>"
... 

bash /udf/deploy_udf.sh
```

Reply:
Yeah agreed, I'll remove it from the file itself